Lesson15/hw_lesson15.py

```diff
@@ -0,0 +1,207 @@
+by input identifiers (type or name). The discount must be specified in percentage
+sell_product(product_name, amount) - removes a particular amount of products from the store if available,
+in other case raises an error. It also increments income if the sell_product method succeeds.
+get_income() - returns amount of many earned by ProductStore instance.
+get_all_products() - returns information about all available products in the store.
+get_product_info(product_name) - returns a tuple with product name and amount of items in the store.
+"""
+
+
+class Product:
+
+    c = 0
+
+    def __init__(self, types: str, name: str, price: float):
+        self.types = types
+        self.name = name
+        self.price = price
+        self.__class__.c += 1
+
+    def __str__(self):
+        return self.types + ', ' + self.name + ', ' + str(self.price)
+
+
+class ProductStore:
+
+    def __init__(self):
+        self.goods = dict()
+
+    def add(self, product, amount):
+        self.goods[product.name] = [product.types, product.name, round(product.price * 1.3, 2), amount]
+        return tuple(self.goods[product.name])
+
+    def set_discount(self, identifier, percent):
+        a = self.goods[identifier]
+        a.append(f'{percent} %')
+        self.goods.update({self.goods.get(self, identifier): a})
+        return self.goods
+
+    def sell(self, product_name, amount):
+        a = self.goods[product_name]
+        a[3] = a[3] - amount
+        self.goods.update({self.goods.get(self, product_name): a})
+        return (product_name, self.goods[product_name])
+
+    @staticmethod
+    def get_income(cls):
+        return cls.c
+
+    def get_all_products(self):
+        for i in self.goods.keys():
+            return i
+
+    def get_product_info(self, product_name):
+        return (product_name, self.goods[product_name][3])
+
+    def __str__(self):
+        return str(self.goods)
+
+
+p = Product('Sport', 'Football T-Shirt', 100)
+
+p2 = Product('Food', 'Ramen', 1.5)
+
+s = ProductStore()
+
+s.add(p, 10)
+
+s.add(p2, 300)
+
+s.sell('Ramen', 10)
+
+assert s.get_product_info('Ramen') == ('Ramen', 290)
+
+
+"""
+Task 4
+Custom exception
+
+Create your custom exception named `CustomException`, you can inherit from base Exception class,
+but extend its functionality to log every error message to a file named `logs.txt`.
+Tips: Use __init__ method to extend functionality for saving messages to file
+"""
+
+
+class CustomException(Exception):
+
+    def __init__(self, msg):
+        self.msg = super().__init__()
+        error_message = msg.__name__
+        text_error = msg.__doc__
+        date_error = datetime.now()
+        with open('../Lesson15/logs.txt', 'a') as file:
+            file.write(f'Date error: {date_error}\n name error: {error_message}\n text error: {text_error}\n')
+            file.write('#' * 100 + '\n')
+
+
+assert CustomException(TypeError)
```

Comment:
The task made correctly, just incorrect usage. Try to use it this way:
```
raise CustomException("message")
```

Reply:
Try, changed, if I understood correctly.